# Переменные и типы данных


<div style="text-align: center;">
  <img src="https://physicsworld.com/wp-content/uploads/2022/03/Planets-in-Alpha-Centauri-system-could-be-carbon-rich-%E2%80%93-if-they-exist.jpg" width="800"/>
</div>



## Обзор

Эта лекция знакомит с базовыми понятиями переменных и типов данных в Python. Также рассматриваются их приложения в астрономии. Это станет основой для последующих тем, связанных с обработкой астрономических данных, так как переменные позволяют хранить и изменять информацию, а типы данных определяют, какие операции можно выполнять с этой информацией. В лекции будут рассмотрены присваивание переменных, правила их наименования и различные типы данных, доступные в Python, с практическими астрономическими примерами.

## Задачи

По завершению этого ноутбука вы научитесь:

- Определять и использовать переменные в Python согласно соглашениям об их именах и присвоении.
- Идентифицировать и использовать различные типы данных Python, включая целые числа, числа с плавающей запятой, строки, булевы переменные, списки и словари.
- Понимать, какие типы данных применяются в астрономическом контексте, таком как координаты и физические характеристики.
- Проделывать базовые операции на различных типах данных.
- Применять понятия переменных в решении реальных задач.

## Что такое переменная?
Переменная — это распространённый термин в программировании для хранения информации. В Python переменная — это символическое имя, которое ссылается на объект или указывает на него. Как только объект присвоен переменной, вы можете обращаться к этому объекту по имени переменной (как "ярлык" для объекта, в котором записаны данные)

Ниже приведён пример создания переменной строкового типа. Кстати, это ближайшая к нам звездная система.

In [1]:
star_name = "Альфа Центавра"

Теперь переменная `star_name` содержит строку `Альфа Центавра`

Чтобы просмотреть содержание переменной, можно воспользоваться функцией `print()`:

In [2]:
print(star_name)

Альфа Центавра


Внутри ячеек Jupyter Notebook можно просто указать имя переменной, и ячейка выведет её содержимое:

In [3]:
star_name

'Альфа Центавра'

## Правила именования переменных

- Имя переменной должно начинаться на букву (любого алфавита, в предпочтении латиница для универсальности) или нижнее подчеркивание `_`.
- В хвосте имени можно использовать буквы, цифры и нижние подчеркивания; дефисы, пробелы и прочие небуквенные символы запрещены.
- Регистр имеет значение: `star_name` и `Star_Name` -- это разные переменные!
- Имя переменной стоит делать содержательным: `star_name`, а не `sn` или `s` -- сами же запутаетесь потом.
- Следующие ключевые слова зарезервированы Python и не могут быть использованы в качестве переменных: `print`, `sum`, `list`, `dict`, `str`, `int`, `float`, `bool`, `set`, `tuple`, `range`, `type`, `object`, `None`, `True`, `False`, `and`, `or`, `not`, `if`, `else`, `elif`, `for`, `while`, `break`, `continue`, `pass`, `def`, `return`, `lambda`, `class`, `import`, `from`, `as`, `with`, `try`, `except`, `finally`, `raise`, `assert`, `del`, `in`, `is`, `global`, `nonlocal`, `yield`, `async`, `await`.

## Типы данных

Типы данных — это различные способы представления информации. Python поддерживает различные типы данных, понимание которых необходимо перед работой с астрономическими данными. Тип данных переменной можно проверять с помощью функции `type()`. Наиболее распространённые типы данных включают:

**а) Целые числа (int):** Этот тип данных представляет целые числа, например, `1, 150, -4`. Они могут быть положительными или отрицательными, но не имеют дробной части (десятичных знаков).

In [13]:
num_features = 20  # Количество параметров в каталоге
type(num_features)

int

**б) Числа с плавающей точкой (float):** Этот тип данных представляет числа с десятичной точкой, например, 3.14, -0.1, 100.5 и т.д. 

В одной ячейке кода можно писать несколько строк кода. Результат будет отображён для последней строки кода.

In [14]:
RA = 219.9021  # Значение прямого восхождения Альфы Центавры
Dec = -60.8340 # Значение склонения Альфы Центавры
type(Dec)

float

**в) Строки (str):** Строки -- наборы символов, таких как "Привет", "Astronomy", "1254", "αβγ".

In [15]:
coordinate_system = "Экваториальная"  # Самая частоиспользуемая координатная система
type(coordinate_system)

str

Для отличия от переменных строки должны быть окаймлены одинарными (`'`) или двойными кавычками (`"`). Можно также использовать тройные кавычки (`'''` or `"""`) для строк из нескольких абзацев:

In [22]:
limerick = """Пожилой господин в Касабланке
Был поклонником жидкой овсянки.
      Чтобы было вкусней,
      В чашку пару мышей
Добавлял господин в Касабланке"""
type(limerick)

str

**г) Булевы переменные (bool):** Булевы переменные могут принимать только два значения: правда (True) или ложь (False)

In [19]:
is_galaxy = True  # Переменная-"флажок" для отличения галактик от не-галактик
type(is_galaxy)

bool

**д) Списки:** Списки -- упорядоченное множество элементов, элементы которого могут быть любым типом данных:

In [31]:
coordinates = [
    219.9021,
    -60.8340,
]  # список с координатами
type(coordinates)

list

**е) Словари (dict):** Словари -- это коллекция из элементов вида ключ-значение:

In [32]:
feature_attributes = {
    "name": "Alpha Centauri A",
    "B_mag": 0.4,
    "type": "Star",
    "coordinates": [219.9021, -60.8340,]
}
type(feature_attributes)

dict

## Служебные символы

Служебные символы используются для вставки символов, которые недопустимы в строке. Например, можно использовать служебный символ `\n` для вставки символа новой строки.

In [24]:
print("Привет!\nЭта строчка разделена.")

Привет!
Эта строчка разделена.


Другой частоиспользуемый служебный символ -- символ табуляции `\t`:

In [26]:
print("Это первая строка.\n\tА это вторая. Смотри, она с отступом!")

Это первая строка.
	А это вторая. Смотри, она с отступом!


Если хочется вставить в строку кавычку, значит, стоит её окаймить другими кавычками или воспользоваться служебным символом `\` для экранирования:

In [29]:
print("Угловые минуты часто обозначаются одним штрихом ', а угловые секунды - двумя штрихами \".")

Угловые минуты часто обозначаются одним штрихом ', а угловые секунды - двумя штрихами ".


## Комментарии

Комментарии используются для объяснения фрагментов кода и служат большей читаемости. В Python комментарии начинаются с символа хэштега `#`. Всё, что находится после `#` игнорируется интерпретатором и не выполняется.

In [33]:
# Вот комментарий на отдельной строке
num_points = 120  # а вот на той же самой, что и код
print(num_points) # вот это выведет значение переменной num_points
# print('А вот эта строчка не выведется)

120


## Операторы

Операторы — это специальные символы или ключевые слова в Python, которые выполняют операции над значениями (называемыми операндами). Считайте их как математические символы `(+, -, *, /)` и логические вентили (and, or, not), которые используются при вычислениях и принятии решений, но адаптированные для программирования. Понимание операторов является основой для выполнения вычислений, сравнения значений и управления ходом решения ваших задач.


### Арифметические операторы
Их используют для вычисления стандартных математических операций.

* `+` : Сложение (выводит сумму двух операндов)
* `-` : Вычитание (выводит разность двух операндов)
* `*` : Умножение (Умножает операнды)
* `/` : Деление (Делит левый операнд на правый, возвращает частное, результат всегда float)
* `//`: Целочисленное деление (Делит, возвращает частное и округляет его вниз до ближайшего целого числа)
* `%` : Остаток от деления (Делит левый операнд на правый и возвращает остаток)
* `**`: Возведение в степень (Возводит левый операнд в степень правого)

### Примеры использования

In [39]:
# расстояния между планетами и Солнцем в астрономических единицах (а.е.)
distance_earth_sun = 1.0
distance_mars_sun = 1.52
distance_jupiter_sun = 5.2

# Сложением найдем расстояние между Землей и Юпитером, если они в противоположных сторонах от Солнца
total_distance = distance_earth_sun + distance_jupiter_sun
print(f"Приблизительное расстояние от Земли до Юпитера: {total_distance} а.е.") # обратите внимание на букву f перед строкой

Приблизительное расстояние от Земли до Юпитера: 6.2 а.е.


In [44]:
# Вычитанием найдем, на сколько Марс дальше от Солнца, чем Земля
distance_difference = distance_mars_sun - distance_earth_sun
print(f"Разность расстояний (Марс - Земля): {distance_difference:.2f} а.е.") # :.Nf после переменной позволяет округлять результат до N знаков

Разность расстояний (Марс - Земля): 0.52 а.е.


In [49]:
# Умножением найдем расстояние от Солнца до Юпитера в километрах (1 а.е. ≈ 150 миллионов км)
distance_jupiter_km = distance_jupiter_sun * 150e6 # 150e6 означает 150, умноженное на 10 в 6 степени
print(f"Расстояние от Юпитера до Солнца: {distance_jupiter_km:.2e} км") # Экспоненциальная запись числа

Расстояние от Юпитера до Солнца: 7.80e+08 км


In [50]:
# Делением найдем, во сколько раз Юпитер дальше от Солнца, чем Марс:
jupiter_mars_ratio = distance_jupiter_sun / distance_mars_sun
print(f"Юпитер в {jupiter_mars_ratio:.2f} раз дальше от Солнца, чем Марс")

Юпитер в 3.42 раз дальше от Солнца, чем Марс


In [51]:
# Деление нацело: положим, АМС летит от Солнца со скоростью 0.5 а.е/год. Сколько полных лет он потратит, летя к Юпитеру?
travel_speed_au_year = 0.5
full_years_to_jupiter = distance_jupiter_sun // travel_speed_au_year
print(f"Чтобы долететь от Солнца до Юпитера на скорости {travel_speed_au_year} а.е./год, потребуется {full_years_to_jupiter} лет")

Чтобы долететь от Солнца до Юпитера на скорости 0.5 а.е./год, потребуется 10.0 лет


In [56]:
# Остаток от деления: сколько а.е. ещё останется спустя ровно 10 лет?
remainder_distance = distance_jupiter_sun % travel_speed_au_year
print(f"Спустя {full_years_to_jupiter} лет АМС нужно будет пролететь ещё {remainder_distance} а.е.")

Спустя 10.0 лет АМС нужно будет пролететь ещё 0.20000000000000018 а.е.


In [58]:
# Возведение в степень: период обращения планеты вокруг Солнца в годах можно оценить, 
# если возвести расстояние от планеты до Солнца в а.е в степень 1.5 (T² ≈ a³)

period_jupyter_year = distance_jupiter_sun ** 1.5
period_mars_year = distance_mars_sun ** 1.5
print(f"Период обращения Юпитера: {period_jupyter_year:.3f} лет")
print(f"Период обращения Марса: {period_mars_year:.3f} лет")

Период обращения Юпитера: 11.858 лет
Период обращения Марса: 1.874 лет


### Операторы сравнения
Эти операторы сравнивают два значения и возвращают True, если проверка выполняется, и False в ином случае:

* `==`: Проверка равенства двух значений (знак двойной для отличия от оператора присваивания `=`)
* `!=`: Проверка неравенства
* `>` : Больше
* `<` : Меньше
* `>=`: Больше или равно
* `<=`: Меньше или равно

Чаще всего операторы сравнения пользуются вместе с **условной конструкцией** `if-else`, позволяющей изменять выполнение кода в зависимости от значения переменной:
```
if (условие, возвращающее True или False):
___%выполняется, если условие True%
else:
___%выполняется, если условие False%
```

Обязательными являются двоеточие `:` в конце условия и отступы в виде табуляции или тройного пробела -- именно ими Python отличает, какой блок кода выполнять в каком случае. 

### Примеры использования операторов сравнения

In [7]:
sun_temp_kelvin = 5778
sirius_temp_kelvin = 9940
earth_mass_relative = 1.0
jupiter_mass_relative = 317.8

# ==: проверка, температура Солнца равна в точности 5800 K?
if (sun_temp_kelvin == 5800):
    print("Температура Солнца в точности равна 5800 K")
else:
    print("Температура Солнца не равна в точности 5800 K")

Температура Солнца не равна в точности 5800 K


In [8]:
# !=: Масса Юпитера не равна массе Земли?
if (jupiter_mass_relative != earth_mass_relative):
    print("Юпитер и Земля имеют разные массы")
else:
    print("Юпитер и Земля имеют одинаковые массы!")

Юпитер и Земля имеют разные массы


In [9]:
# >: Сириус горячее Солнца?
if (sirius_temp_kelvin > sun_temp_kelvin):
    print("Сириус горячее Солнца")
else:
    print("Сириус не горячее Солнца")

Сириус горячее Солнца


In [10]:
# <: Земля менее массивна, чем Юпитер?
if (earth_mass_relative < jupiter_mass_relative):
    print("Земля менее массивна, чем Юпитер")
else:
    print("Земля не менее массивнее Юпитера")

Земля менее массивна, чем Юпитер


In [65]:
# >=: Температура Солнца >= 5500 K?
is_sun_temp_ge_5500 = (sun_temp_kelvin >= 5500)
print(f"Температура Солнца ≥ 5500 K? {is_sun_temp_ge_5500}")

Температура Солнца ≥ 5500 K? True


In [66]:
# Меньше или равно: Масса Земли <= 1?
is_earth_mass_le_1 = (earth_mass_relative <= 1.0)
print(f"Масса Земли <= 1? {is_earth_mass_le_1}")

Масса Земли <= 1? True


### Логические операторы
Эти операторы соединяют несколько условий, преобразуя булевы переменные по заданным правилам.

* `and`: Возвращает True тогда и только тогда, когда оба операнда True.
* `or` : Возвращает False тогда и только тогда, когда оба операнда False.
* `not`: Инвертирует операнд справа (заменяет True на False и False на True).

### Примеры использования логических операторов

In [13]:
# Квазипланеты
planet_distance_au = 1.0 # Расстояние как у Земли
planet_has_liquid_water = True # есть ли жидкая вода? есть
planet_is_gas_giant = False # газовый гигант? да
planet_is_rocky = True # твердая планета? да

# 'and': Планета потенциально обитаема (каменная И с жидкой водой)?
if (planet_is_rocky and planet_has_liquid_water):
    print("Планета потенциально обитаема")
else:
    print("Планета необитаема")

Планета потенциально обитаема


In [70]:
# 'or': Планета далеко от своей звезды ( > 5 а.е.) ИЛИ является газовым гигантом?
is_far_or_gas = (planet_distance_au > 5.0) or planet_is_gas_giant
print(f"Планета далеко от своей звезды ИЛИ является газовым гигантом? {is_far_or_gas}")

Планета далеко от своей звезды ИЛИ является газовым гигантом? False


In [71]:
# 'not': Планета НЕ газовый гигант?
is_not_gas_giant = not planet_is_gas_giant
print(f"Планета НЕ газовый гигант? {is_not_gas_giant}")

Планета НЕ газовый гигант? True


In [72]:
# Сочетаем: планета обитаема И не слишком далеко от своей звезды ( < 1.5 a.e.)?
habitable_and_close = is_potentially_habitable and (planet_distance_au < 1.5)
print(f"Планета обитаема И не слишком далеко от своей звезды? {habitable_and_close}")

Планета обитаема И не слишком далеко от своей звезды? True


### Операторы присваивания
Этими операторами переменным присваиваются конкретные значения. Основной это `=`, однако есть сокращения для сочетания арифметических операторов и операторов присваивания. Особняком стоит моржевой оператор `:=`:

* `=`  : Присвоить значение справа переменной слева.
* `+=` : Добавить правый операнд к левому операнду и присвоить результат левому операнду (например, `x += 1` полностью аналогично `x = x + 1`).
* `-=` : Вычесть правый операнд из левого операнда и присвоить результат левому операнду.
* `*=` : Умножить правый операнд на левый операнд и присвоить результат левому операнду.
* `/=` : Разделить правый операнд на левый операнд и присвоить частное левому операнду.
* `//=`: Разделить правый операнд на левый операнд нацело и присвоить частное левому операнду.
* `%=` : Разделить правый операнд на левый операнд нацело и присвоить остаток левому операнду.
* `**=`: Возвести правый операнд в степень, равную левому операнду, и присвоить результат левому операнду.
* `:=` : Присвоить значение справа переменной слева и **вернуть его как результат операции**.

### Примеры использования

In [73]:
# Положим, мы считаем число планет в Солнечной системе
number_of_planets = 8
print("В солнечной системе", number_of_planets, "планет")

# +=: Положим, что в звездном скоплении из 5 звезд открыли ещё две:
stars_in_cluster = 5
stars_in_cluster += 2
print("Полное число звезд в скоплении:", stars_in_cluster)

В солнечной системе 8 планет
Полное число звезд в скоплении: 7


In [74]:
# -= : Галактика потеряла одну свою спутницу из-за сближения
satellite_galaxies = 4
satellite_galaxies -= 1
print("Оставшихся галактик-спутниц:", satellite_galaxies)

Оставшихся галактик-спутниц: 3


In [75]:
# *= : Расстояние до звезды измерили по новой и оно увеличилось втрое
distance_to_star = 4  # в световых годах
distance_to_star *= 3
print("Новое расстояние до звезды:", distance_to_star, "световых лет")

Новое расстояние до звезды: 12 световых лет


In [45]:
# /= : Полная масса галактики распределена по её секторам
total_mass = 300  # в миллиардах солнечных масс
sectors = 3
mass_per_sector = total_mass
mass_per_sector /= sectors
print("Масса одного галактического сектора:", mass_per_sector, "миллиардов солнечных масс")

Mass per sector: 100.0 billion solar masses


In [76]:
# //= : Распределеим внесолнечные планеты (экзопланеты) поровну среди телескопов для наблюдений
exoplanets = 22
telescopes = 5
exoplanets_per_telescope = exoplanets
exoplanets_per_telescope //= telescopes
print("Каждый телескоп наблюдает", exoplanets_per_telescope, "экзопланеты")

Каждый телескоп наблюдает 4 экзопланеты


In [77]:
# %= : Планета обращается вокруг своей звезды за 365 дней. На сколько дней недели смещается Новый Год каждый год?
orbit_day = 365
orbit_day %= 7  # дней в неделе
print("Новый год сместится на", orbit_day, "дней в неделе")

Новый год сместится на 1 дней в неделе


In [79]:
# **= : Интенсивность света падает обратно пропорционально квадрату расстояния
distance_factor = 2
intensity_factor = distance_factor
intensity_factor **= 2  # квадрат расстояния
print("Интенсивность света упала в", intensity_factor, "раза")

Интенсивность света упала в 4 раза


## Выводы

Понятия переменных и типов данных являются ключевыми в анализе любых данных.
Именно они будут являться фундаментом для вашей дальнейшей работы.
Продолжайте практиковаться, и понимание придет.

## Дополнительные материалы

Для большей информации про типы данных и переменные: 

* (англ) Раздел **Basics** в книге A Byte of Python:  <https://python.swaroopch.com/basics.html>, <https://python.swaroopch.com/op_exp.html> или \
официальной документации <https://docs.python.org/3/tutorial/introduction.html>.
* https://education.yandex.ru/handbook/python, разделы 1 и 2.1


## Вопросы и упражнения

**Переменные**

1.  Как создать переменную `planet_name` и записать в неё строчку `"Юпитер"`?
2.  Как создать переменную `distance_to_sun` и записать в неё число `149.6` (имея в виду миллионы км)?
3.  Если у вас в коде уже есть объявленная переменная `star_temp = 5500`, как изменить её значение на `6000`? 
4.  Как создать переменную `is_gas_giant` и записать в неё `True`, если речь идет про Юпитер?
5.  Если у вас в коде уже есть объявленная переменная `moon_count = 2`, как вывести сообщение "У Марса 2 спутника.", используя эту переменную?
6.  Имя переменной вида `Число Звезд` подходит под стандарты именования в Python? Почему?
7.  Как создать переменную `song` и записать туда в несколько строк первое четверостишье из песни Земляне - Трава у Дома:
   > Земля в иллюминаторе,
> 
   > Земля в иллюминаторе,
> 
   > Земля в иллюминаторе
> 
   > Видна

    Приведите как минимум два способа.

**Операторы**

1.  Земля находится на расстоянии 1.0 а.е. от Солнца, а Марс на расстоянии 1.52 а.е. Как с помощью оператора `-` посчитать разницу между двумя этими расстояниями?
2.  Юпитер обращается вокруг Солнца за 11.86 земных лет. Земной год содержит 365.26 дней. Как сосчитать период обращения Юпитера в днях с помощью Python?
3.  Как увеличить численно значение переменной `jupiter_moons` на `3`? Приведите два разных способа.
4.  Если в переменных `venus_temperature` и `earth_temperature` содержатся температуры Венеры и Земли соответственно, как проверить, какая из них больше? С помощью каких операторов это можно сделать? Проверьте результат, если переменные содержат `462` и `15` градусов, соответсвенно.
5.  Как оператором `==` проверить, равно ли значение переменной `planet_name` в точности строке `"Земля"`?
6.  Планета обитаема, если `has_water == True` И `has_atmosphere == True`. Как использовать оператор `and`, чтобы проверить, выполнены ли оба эти условия?
7. Наблюдения возможны, если `is_night == True` или `is_solar_telescope == False` Как использовать оператор `or`, чтобы проверить, выполнено ли это условие?
8. Проксима Центавра находится на расстоянии 4.24 световых лет (`distance_ly`). Один световой год равен `9.461e12` км (`km_per_ly`). Как вычислить общее расстояние в километрах, записать его в переменную `distance_km` и затем вывести его на экран?